### PR TITLE
Feature/httpcache extensions

### PR DESCRIPTION
**Pull request by Sc0rpic0m**
Added expiry per cache config (not supported in standard memcache) which allows custom expiry time per config
Added caffeine cache store (supports per config expiry) 
Added getStoreType() on HttpCacheStore
Added unit tests to improve coverage

### CODE REVIEW

File: bundle/src/main/java/com/adobe/acs/commons/functions/impl/SupplierWithException.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.adobe.acs.commons.functions.impl;
+
+
+@FunctionalInterface
+public interface SupplierWithException<T> {
```
Comment (badvision) [inline bundle/src/main/java/com/adobe/acs/commons/functions/impl/SupplierWithException.java]: How is this different from CheckedSupplier?

File: bundle/src/main/java/com/adobe/acs/commons/functions/impl/FunctionWithException.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ * <p>
+ * WHAT PURPOSE THAT IT HAS ???
+ * </p>
+ *
+ * @author niek.raaijkmakers@external.cybercon.de
+ * @since 2018-11-24
+ */
+
+@FunctionalInterface
+public interface FunctionWithException<T, R> {
```
Comment (badvision) [inline bundle/src/main/java/com/adobe/acs/commons/functions/impl/FunctionWithException.java]: How does this differ from CheckedFunction?

File: bundle/src/main/java/com/adobe/acs/commons/functions/impl/ConsumerWithException.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * limitations under the License.
+ * #L%
+ */
+package com.adobe.acs.commons.functions.impl;
+
+import java.util.Objects;
+
+
+@FunctionalInterface
+public interface ConsumerWithException<T> {
```
Comment (badvision) [inline bundle/src/main/java/com/adobe/acs/commons/functions/impl/ConsumerWithException.java]: How does this differ from CheckedConsumer?

**Comment by badvision** [comment]
I think that the functions.impl.* classes introduced are redundant next to the Checked* classes in the functions package.  Please let me know if that is not the case.

**Comment by badvision** [comment]
`[ERROR] /home/travis/build/Adobe-Consulting-Services/acs-aem-commons/bundle/src/main/java/com/adobe/acs/commons/functions/CheckedBiFunction.java:[23,44] package com.adobe.acs.commons.functions.impl does not exist`

File: bundle/src/main/java/com/adobe/acs/commons/httpcache/config/impl/HttpCacheConfigImpl.java [hunk 4]
```diff
@@ -175,6 +178,26 @@
...
+            description = "If specified higher then 0, it will overrule the default expiry time with config specific value. Not supported for MEM cache, but supported for all others.",
+            longValue = DEFAULT_EXPIRY_ON_CREATE)
+    private static final String PROP_EXPIRY_ON_CREATE = "httpcache.config.expiry-on-create";
+    private long expiryOnCreate;
+
+    private static final long DEFAULT_EXPIRY_ON_ACCESS = 0L;
+    @Property(label = "Expiry on access",
+            description = "If specified higher then 0, will refresh the expiry time with given value if accessed. Not supported for MEM cache, but supported for all others.",
+            longValue = DEFAULT_EXPIRY_ON_ACCESS)
+    private static final String PROP_EXPIRY_ON_ACCESS = "httpcache.config.expiry-on-access";
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/httpcache/config/impl/HttpCacheConfigImpl.java]: httpcache.config.expiryonaccess

File: bundle/src/main/java/com/adobe/acs/commons/httpcache/store/caffeine/impl/AbstractCaffeineCacheMBean.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    @Override protected final Map<K, V> getCacheAsMap()
+    {
+        return getCache().asMap();
+    }
+
+    @Override
+    @SuppressWarnings("squid:S1192")
+    public final TabularData getCacheStats() throws OpenDataException {
+        // Exposing all google guava stats.
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/httpcache/store/caffeine/impl/AbstractCaffeineCacheMBean.java]: Incorrect commend? This isnt about Guava stats right?

File: bundle/src/test/resources/com/adobe/acs/commons/httpcache/store/jcr/impl/visitor/mock/cachecontent.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<div class="acs-commons">
+    Some awesome cached content!
```
Comment (davidjgonzalez) [inline bundle/src/test/resources/com/adobe/acs/commons/httpcache/store/jcr/impl/visitor/mock/cachecontent.html]: 👍

**Comment by davidjgonzalez** [comment]
@Sc0rpic0m this will need some merge conflict resolution and updating the @Properties to OSGi annotations since the big move

**Comment by Sc0rpic0m** [comment]
I solved the merge conflicts, just have to retest if it still works, standby

**Pull request merged**